### DISCUSSION TITLE
DagFileProcessor running continuously for DAG file (non-UTC system time)

### DISCUSSION THREAD

**Discussion by brki** [apache/airflow#18696] [General]
### Apache Airflow version

2.1.3

### Operating System

Redhat 6.10

### Deployment

Virtualenv installation

### What happened

When deploying an updated version of a DAG, I noticed that there was higher CPU usage, and a new process was starting every few seconds:
```
airflow  31449 76.0  0.1 935636 79236 ?        Rl   17:04   0:00  airflow scheduler - DagFileProcessor /path/to/the/dag_file.py
```

Note: there are many DAGs in this directory, only this file was being continually scanned by the DagFileProcessor.
The other files would be scanned every 5 minutes or so, which agrees with the airflow configuration.

Eventually I guessed that it might be related to a difference between the UTC time and the local system time (Central Europe, e.g. Berlin).

Running 
```
touch -d "1 day ago" /path/to/the/dag_file.py 
```
stopped the continuous scanning of the file.

But after running
```
touch /path/to/the/dag_file.py 
```
the continuous scanning started again.

### What you expected to happen

When a new or updated file appears, I'd expect Airflow to notice, "hey, there's a new or updated file, I'll scan it".   But not scan it again and again.

### How to reproduce

Run Airflow on a system with non-UTC time, specifically 'Europe/Berlin'.

Deploy a new or updated DAG.

Notice that there's a DagFileProcessor process running very often (every second or so there's a new process).

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by nikochiko** [comment]
This issue exists when the last updated timestamp of a file is greater than current time (when I tried `touch -d 'tomorrow' example-dag.py`).

But, I wasn't able to get this issue by setting the system time to another timezone. I am running a docker deployment with airflow 2.1.3

```bash
sudo timedatectl set-timezone Europe/Berlin

touch dags/example-dag.py
docker container top <container-name> | grep DagFileProcessor

# to grab container name, $(docker ps | awk '/scheduler/  { print $NF }')
```

is there something I might have missed?

**Comment by potiuk** [comment]
the problem is that you have likely differnet machines and their time is not synchronized. Timezone setting should have no impact on checking modification time but wroing clock setting likely has. Make sure that all your machines have time synchronized (for example using `ntp`). I am converting that into a discussion as this is not very likely to be an airflow issue.

**Reply by brki** [reply]
Hi @potiuk , thanks for looking into the problem.

It's interesting that you can't reproduce it.

This installation is using `LocalExecutor`; everything is on the same machine.

Possibly related airflow.cfg settings:
```
[core]
default_timezone = utc
executor = LocalExecutor
donot_pickle = False

[scheduler]
min_file_process_interval = 600
min_file_parsing_loop_time = 1
dag_dir_list_interval = 300
parsing_processes = 2
processor_poll_interval = 5


# probably not relevant ...:
[webserver]
default_ui_timezone = Europe/Berlin
```

I consistently see the described behaviour. I tried it again this morning with this simple dag file:
```
import airflow
from airflow.operators.dummy import DummyOperator
from datetime import datetime


with airflow.DAG(
    'debug_start_date_naive',
    catchup=False,
    start_date=datetime(2021, 10, 3),
    schedule_interval='0 * * * *',
    max_active_runs=1,
) as dag:
    dummy_task = DummyOperator(task_id='dummy')
```

System time / time zone:
```
$ date
Mon Oct  4 11:46:10 CEST 2021

$ date +"%Z %z"
CEST +0200
```

**Comment by potiuk** [comment]
My guess is that the timezone change that you do is not what you think it is. 

You can easily check it by running `date -u`  before and after the change - they should show the same time (+/- the few seconds between the checks). My guess is when you change the time zone, you also change the UTC clock of the computer - if UTC time also changes when you change the time zone case this means that your way of changing the timezone is broken.

**Reply by brki** [reply]
I'm not changing the timezone at all actually.  It's already set, it's on a managed production server.

```
$ date; date -u
Mon Oct  4 14:28:50 CEST 2021
Mon Oct  4 12:28:50 UTC 2021
```

**Reply by potiuk** [reply]
I was under the impression that you actually change it.

Then you likely have a problem with the process of DAG deployment. Likely - whatever the process you are using, creates the DAG with mtime or ctime in the future. I do not know how you deploy the DAG but you should check modifications times of the dags. Maybe you are using some filesystem that is wrongly configured and skews the time - you should lool what kind of deployment you have and how you deploy the DAGs and how the process changes various times (creation/modification/access) of your files

**Comment by brki** [comment]
Hmm ... I don't think that's the problem either.  Otherwise touching the file wouldn't affect the behaviour the way it does.

If a file is `touch`ed without arguments (e.g. the file access and modification time is set to the current time), DagFileProcessor starts reprocessing it constantly.

If a file that is constantly being processed by DagFileProcessor is `touch`ed with an argument to put the access + modification time before the UTC time difference (e.g. currently two hours for CEST), the DagFileProcessor stops constanstly processing the file.

**Comment by brki** [comment]
Ah, looking at the Airflow code I think I see what is causing this behavior.

https://github.com/apache/airflow/blob/2.1.4/airflow/dag_processing/manager.py#L1056

```
file_modified_time = timezone.make_aware(datetime.fromtimestamp(files_with_mtime[file_path]))
```

This results in the naiive datetime being converted to a UTC datetime (because of my `core.default_timezone = utc` configuration).

Perhaps I should just set my core.default_timezone to 'Europe/Berlin' ?  Is that safe to do nowadays?

We started using Airflow around version 1.6 or 1.8.  At that time I think it was not recommended to use a non-UTC timezone.

**Reply by ibr** [reply]
Setting core.default_timezone to 'Europe/Berlin' fixed that problem for me. Thanks for investigating.
